ouroboros-consensus/test/infra-test/Test/Ouroboros/Consensus/ChainGenerator/Tests/BlockTree.hs

```diff
@@ -75,3 +88,84 @@ slotLength fragment =
   fromIntegral $ unSlotNo $
     withOrigin 0 id (AF.anchorToSlotNo $ AF.headAnchor fragment)
     - withOrigin 0 id (AF.anchorToSlotNo $ AF.anchor fragment)
+
+-- | Return all the full fragments from the root of the tree.
+allFragments :: BlockTree blk -> [AF.AnchoredFragment blk]
+allFragments BlockTree{..} = trunk : map full branches
+
+firstJust :: [Maybe a] -> Maybe a
+firstJust [] = Nothing
+firstJust (Nothing : rest) = firstJust rest
+firstJust (Just x : _) = Just x
+
+-- | Look for a point in the block tree and return a fragment going from the
+-- root of the tree to the point in question.
+findFragment :: AF.HasHeader blk => AF.Point blk -> BlockTree blk -> Maybe (AF.AnchoredFragment blk)
+findFragment point blockTree =
+  fst <$> firstJust (map (\fragment -> AF.splitAfterPoint fragment point) $ allFragments blockTree)
+
+newtype InvolvesRollback = InvolvesRollback Bool
+
+findPath :: AF.HasHeader blk => AF.Point blk -> AF.Point blk -> BlockTree blk -> Maybe (InvolvesRollback, AF.AnchoredFragment blk)
+findPath source target blockTree = do
+  sourceFragment <- findFragment source blockTree
+  targetFragment <- findFragment target blockTree
+  (_, _, sourceSuffix, targetSuffix) <- AF.intersect sourceFragment targetFragment
+  pure (InvolvesRollback (AF.length sourceSuffix /= 0), targetSuffix)
+
+prettyPrint :: AF.HasHeader blk => BlockTree blk -> [String]
```

Comment:
Having a small example of the output here would help reading the code.

Reply:
Done in bc542836d but I think I should also just rework this function to be more generic and also to make it much easier to read.